File: go/384.go
/*

https://leetcode.com/problems/shuffle-an-array/

Shuffle a set of numbers without duplicates.

Example:

// Init an array with set 1, 2, and 3.
int[] nums = {1,2,3};
Solution solution = new Solution(nums);

// Shuffle the array [1,2,3] and return its result. Any permutation of [1,2,3] must equally likely to be returned.
solution.shuffle();

// Resets the array back to its original configuration [1,2,3].
solution.reset();

// Returns the random shuffling of array [1,2,3].
solution.shuffle();

*/

package main


import "fmt"
import "math/rand"
import "time"


type Solution struct {
    Array []int
}


func Constructor(nums []int) Solution {
    obj := Solution{nums}
    return obj
}


/** Resets the array to its original configuration and return it. */
func (this *Solution) Reset() []int {
    return this.Array 
}


/** Returns a random shuffling of the array. */
func (this *Solution) Shuffle() []int {
    buf := make([]int, len(this.Array))
    copy(buf, this.Array)
    rand.Seed(time.Now().UnixNano()) 
    rand.Shuffle(len(buf), func(i, j int) {buf[i], buf[j] = buf[j], buf[i]})
    return buf
}


func main() {
    nums := []int{3,2, 1}    
    s := Constructor(nums)
    fmt.Println(s.Reset())
    fmt.Println(s.Shuffle())
    fmt.Println(s.Shuffle())
    fmt.Println(s.Reset())
}

File: python/384.py
"""
https://leetcode.com/problems/shuffle-an-array/

Shuffle a set of numbers without duplicates.

Example:

// Init an array with set 1, 2, and 3.
int[] nums = {1,2,3};
Solution solution = new Solution(nums);

// Shuffle the array [1,2,3] and return its result. Any permutation of [1,2,3] must equally likely to be returned.
solution.shuffle();

// Resets the array back to its original configuration [1,2,3].
solution.reset();

// Returns the random shuffling of array [1,2,3].
solution.shuffle();

"""

from typing import List


import random

class Solution:

    def __init__(self, nums: List[int]):
        self.buf = nums

    def reset(self) -> List[int]:
        """
        Resets the array to its original configuration and return it.
        """
        return self.buf


    def shuffle(self) -> List[int]:
        """
        Returns a random shuffling of the array.
        """
        tmp = self.buf[:]
        random.shuffle(tmp)
        return tmp


# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()
